pkg/pods/pod.go

```diff
@@ -466,17 +472,30 @@ func (pod *Pod) Install(manifest manifest.Manifest, verifier auth.ArtifactVerifi
 			continue
 		}
 
-		launchableURL, verificationData, err := artifactRegistry.LocationDataForLaunchable(pod.Id, launchableID, stanza)
-		if err != nil {
-			pod.logLaunchableError(launchable.ServiceID(), err, "Unable to install launchable")
-			return err
-		}
+		// TODO: make this code better, probably abstract away launchable installation
+		// into something that understands the types
+		if launchable.Type() == "hoist" || launchable.Type() == "opencontainer" {
+			launchableURL, verificationData, err := artifactRegistry.LocationDataForLaunchable(pod.Id, launchableID, stanza)
+			if err != nil {
+				pod.logLaunchableError(launchable.ServiceID(), err, "Unable to install launchable")
+				return err
+			}
 
-		err = downloader.Download(launchableURL, verificationData, launchable.InstallDir(), manifest.UnpackAsUser())
-		if err != nil {
-			pod.logLaunchableError(launchable.ServiceID(), err, "Unable to install launchable")
-			_ = os.Remove(launchable.InstallDir())
-			return err
+			err = downloader.Download(launchableURL, verificationData, launchable.InstallDir(), manifest.UnpackAsUser())
+			if err != nil {
+				pod.logLaunchableError(launchable.ServiceID(), err, "Unable to install launchable")
+				_ = os.Remove(launchable.InstallDir())
+				return err
+			}
+		} else if launchable.Type() == "docker" {
+			// TODO: auth?
+			resp, err := pod.DockerClient.ImagePull(context.TODO(), stanza.Image.Name, dockertypes.ImagePullOptions{})
+			if err != nil {
+				pod.logLaunchableError(launchable.ServiceID(), err, fmt.Sprintf("could not pull docker image: %s", err))
+				return util.Errorf("could not pull docker image: %s", err)
+			}
+			// TODO: anything we want in this response body?
```

Comment:
The Docker client returns an HTTP response body and expects the caller to parse it? What kind of an API is that?!

That (lack of) design has implications here: `err` checks for network errors and `resp.StatusCode` checks whether the operation succeeded or failed. From the REST [API reference](https://docs.docker.com/engine/api/v1.30/#operation/ImageCreate), if the call fails, the response body contains the error message. It looks like no response body is returned on success.

Reply:
yeah definitely a weird API. looking through their code it does look like they return an error from `ImagePull` if the response code is `< 200` or `>= 400`, and that error includes the response body. So I don't think there's ever anything useful in the response?